20250404-RS-workflow-rerun-from-activity.md

```diff
@@ -0,0 +1,142 @@
+# Workflow: Rerun from Activity
+
+* Author(s): @joshvanl, @whitwaldo
+
+## Overview
+
+This proposal details the ability to rerun a workflow from a previous point in its history.
+A workflow in a terminal state can be rerun from a failed activity, before the failed activity, or at any activity in the history of a successful or failed workflow.
+
+## Background
+
+It is often the case that it's desirable to re-run business logic implemented inside a workflow.
+This could be because an activity in the workflow failed due to a transient error, an external dependency changed or a resource is now available to an activity, or it's just desirable for some subset of the last set of Activities to be rerun.
+Dapr should provide the functionality to rerun a workflow from any point in its history.
+
+## Related Items
+
+https://github.com/dapr/proposals/pull/79
+
+## Design
+
+The following proto RPC and messages will be added to durabletask, exposed via each SDK.
+This API implements rerunning a workflow from a specific Activity in the history of the workflow.
+The Activity to be rerun is chosen via its associated event ID.
+
+All `Activties` are assigned an event ID in the durabletask history.
+While `Timers` and `RaiseEvents` tasks are also assigned an event ID in the durabletask history, the workflow cannot be rerun from these events using this API.
+Not only would supporting rerunning the workflow from these two event types require a significant code refactor, in practice, users are only interested in rerunning workflow from a specific _activity_, not a "control event".
+
+It must be the case that the workflow is in a _terminal_ state before the rerun can be executed.
+This would be because the workflow has completed successfully, failed at some activity, or force terminated.
+Rerunning a workflow which is currently in progress does not make any practical or academic sense.
+Attempting to do so will return an error to the client.
+
+When rerunning a workflow, the workflow will be started from the event ID of an Activity in the history.
+The workflow actor will delete all history events after the event ID of the Activity chosen.
```

Comment:
Does this mean we'd lose the history of the original execution - is possible to retain/archive that?

Reply:
Ah think you answered this in a later point

Reply:
This is where I prefer the checkpoint -> clone event state because this approach makes the workflows themselves mutable and subject to change by post-terminal operations. Rather than re-run a workflow and modify the events that potentially have run (even if the larger workflow didn't), it feels icky to modify the workflow data itself instead of clone it out to a distinctly separate workflow subject to its own state.

Reply:
I think we could just change this from optional to required
```
The client can optionally give a new instance ID to use when rerunning the workflow.
```
then it won't be possible to mutate the workflow history post-execution

Reply:
I disagree - this proposal does not call for cloning the workflow state as mine does, but explicitly mutates the workflow history. Because there's no cloning, I don't know how your new workflow would have access to the event history of another workflow.

In my own, because you're using a cloned history from some intermediate step in another workflow, you're never mutating anything.

Reply:
I’ll be updating the proposal to clone the state.